### ISSUE TITLE
Enable Symfony DI autowiring

### ISSUE THREAD

**Issue by vincentchalamon** [Behat/Symfony2Extension#128]
As a bridge with Symfony2, it should be interesting to autowire contexts arguments using Symfony DI autowiring, if they're detected as Symfony services. For instance:
```yaml
# behat.yml
default:
    suites:
        default:
            contexts:
                - FeatureContext: ~
    extensions:
        Behat\Symfony2Extension:
            kernel:
                bootstrap: 'vendor/autoload.php'
                path: 'src/Kernel.php'
                class: 'App\Kernel'
```
```php
class FeatureContext implements Context
{
    private $clientManager;
    private $tokenManager;

    public function __construct(ClientManagerInterface $clientManager, TokenManagerInterface $tokenManager)
    {
        $this->clientManager = $clientManager;
        $this->tokenManager = $tokenManager;
    }

    // ...
}
```

**Comment by sroze** [comment]
Definitely agree with you @vincentchalamon. Actually, we might just rewrite a Behat extension for Symfony 4+...

**Comment by vincentchalamon** [comment]
A different extension for Symfony2 vs Symfony4? Why no retro-compatibility?

**Comment by sroze** [comment]
Something like that :)

**Comment by vincentchalamon** [comment]
@sroze As I've seen, this extension loads the container after kernel boot (which is normal). At this point, I cannot access any ContainerBuilder to look for a service by its class name. Do you know another way to do it?

**Comment by tsantos84** [comment]
Same question as @vincentchalamon here. If I understood correctly the process of context creation and argument resolving, `Symfony2Extension` should be able to share its container definitions with kernel container and create the Context classes as a service. By doing this `Behat` will create the contexts from service container and its arguments will be resolved directly by the container auto-wiring mechanism.

**Comment by sroze** [comment]
Erm, that's a good idea @tsantos84

**Comment by tsantos84** [comment]
A workaround to inject your private services to contexts may be create a [CompilerPass](https://api.symfony.com/4.0/Symfony/Component/DependencyInjection/Compiler/CompilerPassInterface.html) to make all service definitions public and register it only on `test`environments. Didn't test this here!

**Comment by sroze** [comment]
I had a quick look and.. even if you idea is very good @tsantos84... @vincentchalamon is right, we do not have the container builder when locating the context (actually's it's behat's job, not from the extension..). Looking at it still 🤔

**Comment by vincentchalamon** [comment]
@tsantos84 To be clear, your idea is to connect the Behat contexts in the Symfony application container through Symfony2Extension? If so, it could unblock lot of features to extend Behat!

Otherwise, as the Behat container is different from the application container, it's still possible to guess the context arguments in the `ServiceArgumentResolver::resolveArguments` method. But we don't have access to the application container builder to guess the service based on the typehint of the context arguments.

**Comment by vincentchalamon** [comment]
Or did you mean to inject the application services in the Behat container? So we could have an access to a container builder on Behat build, and it _could_ allow to guess a service if by its typehint

**Comment by tsantos84** [comment]
Exactly @vincentchalamon, [ContainerBuilder](http://api.symfony.com/4.0/Symfony/Component/DependencyInjection/ContainerBuilder.html) has a [method](http://api.symfony.com/4.0/Symfony/Component/DependencyInjection/ContainerBuilder.html#method_merge) to merge another container. So, if we merge behat with application containers may we could register the context classes as a service and use the container to instantiate them.

**Comment by stof** [comment]
@tsantos84 this would merge service definitions, but you would still have 2 separate containers: the one used by Behat, and the one used by the kernel. so you would not reuse the same service instances.

**Comment by vincentchalamon** [comment]
@stof 👍
But I think this approach could still help to guess the service id by it's typehint from the context arguments (in the container builder process in Behat). Then we could get the right service from the application container.
Also, merging all services in the Behat container _may be_ not required as we won't use thoses services after build.

**Comment by tsantos84** [comment]
@stof do we need the kernel container definitions once it was merged to behat's container?

**Comment by tsantos84** [comment]
@vincentchalamon we still have problem with auto-wiring private services, right?

**Comment by sroze** [comment]
> we still have problem with auto-wiring private services, right?

Correct, but you'd have Symfony DI's error message that is quite good. i.e. something like "the service is not accessible because private, make it public, blah blah"

**Comment by tsantos84** [comment]
I agree that this approach is a way to follow, but we would loose the performance improvement added recently to DI component by making services public just to fit the `Behat` architecture.

**Comment by tsantos84** [comment]
@stof I see your point and you're right. We still have two redundant containers with reduntant services. That said, what about to create a Behat application with an existing `ContainerBuilder`?

**Comment by vincentchalamon** [comment]
@sroze Did you mean something like that? 😄 
<img width="1274" alt="capture d ecran 2017-12-20 a 09 05 59" src="https://user-images.githubusercontent.com/407859/34197039-088c0528-e565-11e7-8710-ccd2ece4bf27.png">

**Comment by sroze** [comment]
What do you mean Vincent?

**Comment by vincentchalamon** [comment]
When you said:
> Correct, but you'd have Symfony DI's error message that is quite good. i.e. something like "the service is not accessible because private, make it public, blah blah"

**Comment by tsantos84** [comment]
Any news about this issue?

**Comment by tsantos84** [comment]
@sroze sure. Did we concluded any final solution in this issue?

**Comment by jcornide** [comment]
hi,

is there any way right now to inject a Symfony 4 service to a behat context without having to touch the compiler pass or similar?

**Comment by warslett** [comment]
@jcornide basically no unless you declare everything as public which means you lose the speed enhancements of the new container. The way I'm doing it at the moment is that I have a few basic services that Behat Contexts interact with directly and they are declared public. So rather than injecting the entity manager into my Behat context I have services, for example TestFixtureService which is public and the entity manager is injected into that with autoloading. I only declare those services in my services_test.yml so they don't get loaded at all in production. It's actually quite a nice design that works really well but it's not necessarily what I would have done if I'd just been able to inject my services as normal.

**Comment by stof** [comment]
> So rather than injecting the entity manager into my Behat context I have services, for example TestFixtureService which is public and the entity manager is injected into that with autoloading

@warslett that's a bad example, as the Doctrine entity_manager is already a public service :smile:

**Comment by DonCallisto** [comment]
Are you guys aware of [this](https://github.com/Behat/Behat/pull/974)? We should only provide access to symfony container in order to leverage on autowire that's now implemented in behat.

**Comment by ciaranmcnulty** [comment]
@DonCallisto yes and I'm using that directly in one codebase by giving it the Symfony container - however it doesn't support parameters.

**Comment by DonCallisto** [comment]
@ciaranmcnulty how do you managed it to work if this extension does not provide that facility?

**Comment by vincentchalamon** [comment]
@DonCallisto but what about private services in the Symfony container?

**Comment by DonCallisto** [comment]
Private ones can be (should be) fetched from a particular container reserved for tests. It's already enabled since symfony 4.1 IIRC. For services never injected, you should alias them with `public: false`

**Comment by DonCallisto** [comment]
@vincentchalamon Take a look [https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing](https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing)

**Comment by vincentchalamon** [comment]
Didn't know about that, it helps me so much in my Behat extension. Thanks for sharing :heart:

**Comment by tyx** [comment]
Hi guys,

anyone start something about that ? even a gist ?

**Comment by tyx** [comment]
Ok I guess everyone moved to https://github.com/FriendsOfBehat/SymfonyExtension and nobody tells me 😭

**Comment by DonCallisto** [comment]
I was just suggesting this and then you found out it by yourself :smile: 
Right now we're not still moved to FOB/SymfonyExtension but we have planned to do this ASAP.

**Comment by sroze** [comment]
How did they solve the issue in FOB/SymfonyExtension?

**Comment by tyx** [comment]
They inversed the principle. You load context directly from your Symfony app. So your contexts files are basically just classic services from Symfony POV with autowire

**Comment by sroze** [comment]
Indeed, much better idea 👍

**Comment by vincentchalamon** [comment]
Closing in favor of https://github.com/FriendsOfBehat/SymfonyExtension
